src/protocol/context/malicious.rs

```diff
@@ -58,17 +71,47 @@ impl<'a, F: Field> MaliciousContext<'a, F> {
     pub(super) fn new<S: Substep + ?Sized>(
         source: &SemiHonestContext<'a>,
         malicious_step: &S,
-        upgrade_ctx: SemiHonestContext<'a>,
         acc: MaliciousValidatorAccumulator<F>,
         r_share: Replicated<F>,
     ) -> Self {
         Self {
-            inner: ContextInner::new(upgrade_ctx, acc, r_share),
+            inner: ContextInner::new(source, acc, r_share),
             step: source.step().narrow(malicious_step),
             total_records: TotalRecords::Unspecified,
         }
     }
 
+    /// TODO: This is not fast, but we can't just `reinterpret_cast` here.
+    fn as_semi_honest(&self) -> SemiHonestContext<'a> {
```

Comment:
This really isn't easy to fix.  It might be that what we really need to write a sparse multiply op that is passed the gateway and PRSS directly so that we can use that for upgrading.